### PR TITLE
Player stats and information

### PR DESCRIPTION
**Pull request by djetelina**
What I forgot but will add tomorrow/later (or someone else can):  Probably a good idea to include number of games and number of wins.

I'm expecting a review of the query, as I'm sure it's suboptimal and more demanding/slower than possible. I'm also totally up for not including some of those stats, but I figured I might as well pull everything (except judgement in progress) and then decide what to keep rather than the opposite.

### CODE REVIEW

File: cbsapi/api_blueprint.py [hunk 1]
```diff
@@ -69,3 +70,97 @@ def ranking():
...
+                                     CAST(SUM(ct.rarity = 1) AS UNSIGNED) AS uncommons,
+                                     CAST(SUM(ct.rarity = 2) AS UNSIGNED) AS rares
+                              FROM cards c
+                                     INNER JOIN card_types ct ON c.type_id = ct.id
+                              GROUP BY owner_id) AS cc ON cc.owner_id = p.id
+            WHERE p.name = %s;""",
+            name
+        )
+        player_data = cursor.fetchone()
+        transformed_data = dict(
```
Comment (djetelina) [inline cbsapi/api_blueprint.py]: This doesn't have to be inside the with statement.

**Comment by kbasten** [comment]
Nice commit, but please fix the card select query. It currently retrieves the stats for every player in the database, and then selects the requested player. That is a pretty heavy operation. Something like this should give you a starting point on how to improve it:

```mysql
SELECT A.rarity, SUM(A.cnt) AS cnt
FROM (
  SELECT CT.rarity, COUNT(*) AS cnt
  FROM cards C, card_types CT
  WHERE C.owner_id = 50001
  AND C.type_id = CT.id
  GROUP BY C.type_id
) AS A
GROUP BY A.rarity
```

**Comment by djetelina** [comment]
I managed to edit my query a little bit and got quite an improvment, the bottom part of it was edited to this (which I don't know why I didn't do in the first place):

```sql
        FROM profiles p
               INNER JOIN profile_data pd ON pd.profile_id = p.id
               LEFT JOIN avatars a ON a.profile_id = p.id
               LEFT JOIN cards c ON c.owner_id = p.id
               LEFT JOIN card_types ct ON ct.id = c.type_id
        WHERE p.name = 'Robot'
        GROUP BY c.owner_id
```

Explain on the queries:

From:

id | select_type | table | partitions | type | possible_keys | key | key_len | ref | rows | filtered | Extra
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | PRIMARY | p | NULL | const | PRIMARY,name | name | 767 | const | 1 | 100 | NULL
1 | PRIMARY | pd | NULL | const | PRIMARY | PRIMARY | 4 | const | 1 | 100 | NULL
1 | PRIMARY | a | NULL | const | PRIMARY | PRIMARY | 4 | const | 0 | 0 | unique row not found
1 | PRIMARY | <derived5> | NULL | ref | <auto_key0> | <auto_key0> | 4 | const | 10 | 100 | Using where
5 | DERIVED | c | NULL | ALL | owner_id | NULL | NULL | NULL | 19302 | 100 | Using temporary; Using filesort
5 | DERIVED | ct | NULL | eq_ref | PRIMARY | PRIMARY | 2 | scrolls.c.type_id | 1 | 100 | NULL
4 | DEPENDENT SUBQUERY | idol_unlocks | NULL | ref | PRIMARY | PRIMARY | 4 | const | 1 | 100 | Using index
3 | DEPENDENT SUBQUERY | avatar_unlocks | NULL | ref | PRIMARY | PRIMARY | 4 | const | 1 | 100 | Using index
2 | DEPENDENT SUBQUERY | achievement_unlocks | NULL | ref | PRIMARY | PRIMARY | 4 | const | 1 | 100 | Using index



To:

id | select_type | table | partitions | type | possible_keys | key | key_len | ref | rows | filtered | Extra
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | PRIMARY | p | NULL | const | PRIMARY,name | name | 767 | const | 1 | 100 | Using temporary; Using filesort
1 | PRIMARY | pd | NULL | const | PRIMARY | PRIMARY | 4 | const | 1 | 100 | NULL
1 | PRIMARY | a | NULL | const | PRIMARY | PRIMARY | 4 | const | 0 | 0 | unique row not found
1 | PRIMARY | c | NULL | ref | owner_id | owner_id | 4 | const | 1885 | 100 | Using where
1 | PRIMARY | ct | NULL | eq_ref | PRIMARY | PRIMARY | 2 | scrolls.c.type_id | 1 | 100 | NULL
4 | DEPENDENT SUBQUERY | idol_unlocks | NULL | ref | PRIMARY | PRIMARY | 4 | const | 1 | 100 | Using index
3 | DEPENDENT SUBQUERY | avatar_unlocks | NULL | ref | PRIMARY | PRIMARY | 4 | const | 1 | 100 | Using index
2 | DEPENDENT SUBQUERY | achievement_unlocks | NULL | ref | PRIMARY | PRIMARY | 4 | const | 1 | 100 | Using index

This got rid of the derived table and now is just joining them and grouping at the end. Here's the difference locally:

`1 row retrieved starting from 1 in 104 ms (execution: 69 ms, fetching: 35 ms)`

`1 row retrieved starting from 1 in 91 ms (execution: 24 ms, fetching: 67 ms)`

Executed twice as fast now. Would that work, or should I keep trying to improve it? :)

**Comment by djetelina** [comment]
Aaand the group by is useless, so no more using temporary using filesort :)

**Comment by LordRalex** [comment]
```sql
SELECT
	p.name AS name,
	pd.alpha_ranked * pd.real_ranked AS rating,
	pd.gold AS gold,
	p.login_date AS last_login,
	p.created AS created,
	au.achievements_unlocked AS achievements_unlocked,
	av_u.avatar_unlocks AS avatar_unlocks,
	i_u.idols_unlocked AS idols_unlocked,
	a.head AS avatar_head,
	a.body AS avatar_body,
	a.leg AS avatar_legs,
	a.arm_back AS arm_back,
	a.arm_front AS arm_front,
	c.commons,
	c.uncommons,
	c.rares,
	gps.games_won,
	gps.games_lost
FROM profiles p
INNER JOIN profile_data pd
	ON pd.profile_id = p.id
LEFT JOIN avatars a
	ON a.profile_id = p.id
LEFT JOIN (SELECT
		owner_id,
		SUM(ct.rarity = 0) AS commons,
		SUM(ct.rarity = 1) AS uncommons,
		SUM(ct.rarity = 2) AS rares
	FROM cards c
	JOIN card_types ct
		ON ct.id = c.type_id
	GROUP BY owner_id) c
	ON c.owner_id = p.id
LEFT JOIN (SELECT
		profile_id,
		SUM(win = 1) AS games_won,
		SUM(win = 0) AS games_lost
	FROM game_player_stats
	GROUP BY profile_id) AS gps
	ON gps.profile_id = p.id
LEFT JOIN (SELECT
		profile_id,
		COUNT(*) AS achievements_unlocked
	FROM achievement_unlocks
	GROUP BY profile_id) au
	ON au.profile_id = p.id
LEFT JOIN (SELECT
		profile_id,
		COUNT(*) AS avatar_unlocks
	FROM avatar_unlocks
	GROUP BY profile_id) av_u
	ON av_u.profile_id = p.id
LEFT JOIN (SELECT
		profile_id,
		COUNT(*) AS idols_unlocked
	FROM idol_unlocks
	GROUP BY profile_id) AS i_u
	ON i_u.profile_id = p.id
/*WHERE p.name = 'DeckTemplate'*/
```

This is a more improved query that actually fixes the counters, but it's not complete and not tested.

**Pull request merged**